core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -2945,6 +2947,34 @@ class KafkaApis(val requestChannel: RequestChannel,
     }
   }
 
+  def handleBrokerHeartbeat(request: RequestChannel.Request): Unit = {
+    val brokerHeartbeatRequest = request.body[BrokerHeartbeatRequest]
+
+    if (authorize(request.context, CLUSTER_ACTION, CLUSTER, CLUSTER_NAME)) {
+      sendResponseMaybeThrottle(request, requestThrottleMs =>
+        brokerHeartbeatRequest.getErrorResponse(requestThrottleMs,
+          Errors.UNSUPPORTED_VERSION.exception))
+    } else {
+      sendResponseMaybeThrottle(request, requestThrottleMs =>
+        brokerHeartbeatRequest.getErrorResponse(requestThrottleMs,
+          Errors.CLUSTER_AUTHORIZATION_FAILED.exception))
+    }
+  }
+
+  def handleControllerHeartbeat(request: RequestChannel.Request): Unit = {
+    val controllerHeartbeatRequest = request.body[ControllerHeartbeatRequest]
+
+    if (authorize(request.context, CLUSTER_ACTION, CLUSTER, CLUSTER_NAME)) {
```

Comment:
It really motivates me to think whether it makes sense to get a separate acl operation such like `BROKER_ACTION`. I couldn't think of a reason one admin client would send a hb request.

Reply:
I guess the question is whether you'd want the controllers to hold separate and different ACLs from what brokers hold.  Let's think about that outside the context of this PR.